File: Analysis and Design of Algorithms/Practice 2/dfs.cpp
#include <iostream>
#include <conio.h>
#include <list>
#include <vector>

using namespace std;

class Graph
{
    int V;
    list<int> *l;

public:
    Graph(int v)
    {
        V = v;
        l = new list<int>[V];
    }

    void addEdge(int u, int v, bool undir = true)
    {
        l[u].push_back(v);
        if (undir)
            l[v].push_back(u);
    }

    void dfshelper(int src, vector<bool> &visited, vector<int> &ans)
    {
        visited[src] = true;
        ans.push_back(src);

        for (int nbr : l[src])
            if (!visited[nbr])
                dfshelper(nbr, visited, ans);
    }

    void dfs(int src)
    {
        vector<int> ans;
        vector<bool> visited(V, false);

        dfshelper(src, visited, ans);

        for (int nbr : ans)
            cout << nbr << ", ";
    }
};

int main()
{
    Graph g(7);

    g.addEdge(1, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 4);
    g.addEdge(2, 5);
    g.addEdge(3, 5);
    g.addEdge(4, 5);
    g.addEdge(4, 6);
    g.addEdge(5, 6);

    cout << "Traversal order is: ";

    // cout << "\nUsing BFS => ";
    // g.bfs(1);

    cout << "\nUsing DFS =>";
    g.dfs(1);

    getch();
    return 0;
}

File: Artificial Intelligence/Practice/dfs.py
class graph:
    vertices = {}

    def __init__(self):
        pass

    def addEdge(self, u, v):
        if u in self.vertices:
            self.vertices[u].append(v)
        else:
            l = list()
            l.append(v)
            self.vertices[u] = l

    def print_graph(self):
        for i in self.vertices:
            print(f"{i} -> {self.vertices[i]}")

    def dfs_helper(self, s, v):
        v.add(s)
        print(s, end=" ")

        try:
            for n in self.vertices[s]:
                if(n not in v):
                    self.dfs_helper(n, v)
        except:
            return

    def dfs(self, source):
        visited = set()
        self.dfs_helper(source, visited)


g = graph()
g.addEdge(0, 1)
g.addEdge(0, 2)
# g.addEdge(1, 2)
g.addEdge(2, 0)
g.addEdge(2, 3)
g.addEdge(3, 3)

g.print_graph()

print("DFS Traversal: ", end="")
g.dfs(2)
